File: code/chap04/python/average_by_key_use_combinebykey.py
#!/usr/bin/python
#-----------------------------------------------------
# This program find average per key using 
# the combineByKey() transformation.
#------------------------------------------------------
# Input Parameters:
#    none
#-------------------------------------------------------
# @author Mahmoud Parsian
#-------------------------------------------------------
from __future__ import print_function
import sys
from pyspark.sql import SparkSession

def create_spark_session():
    spark = SparkSession\
        .builder\
        .appName("average_by_key_use_aggregatebykey")\
        .getOrCreate()
    return spark
#end-def

def main():

    # create an instance of SparkSession
    spark = create_spark_session()

    input = [("k1", 1), ("k1", 2), ("k1", 3), ("k1", 4), ("k1", 5),\
             ("k2", 6), ("k2", 7), ("k2", 8),\
             ("k3", 10), ("k3", 12)]

    # build RDD<key, value>
    rdd = spark.sparkContext.parallelize(input)

    # Combined data structure (C) is a Tuple2(sum, count)
    # 3 functions needs to be defined:
    #    v --> C
    #    C, v --> C
    #    C, C --> C
    sum_count = rdd.combineByKey(\
        lambda v: (v, 1),\
        lambda C, v: (C[0] + v, C[1] + 1),\
        lambda C1, C2: (C1[0] + C2[0], C1[1] + C2[1])\
    )

    # show sum_count
    print("sum_count = ", sum_count.collect())
    # [
    #  ('k3', (22, 2)), 
    #  ('k2', (21, 3)), 
    #  ('k1', (15, 5))
    # ]

    # find averages 
    avg = sum_count.mapValues( lambda v : float(v[0]) / float(v[1]))
    print("avg = ", avg.collect())
    # avg.collect()
    # [
    #  ('k3', 11), 
    #  ('k2', 7), 
    #  ('k1', 3)
    # ]
    
    # done!
    spark.stop()
#end-def

if __name__ == '__main__':
    main()

File: code/chap04/scala/src/main/scala/org/data/algorithms/spark/ch04/AverageByKeyUseCombineByKey.scala
package org.data.algorithms.spark.ch04

import org.apache.spark.sql.SparkSession

object AverageByKeyUseCombineByKey {
  def createSparkSession: SparkSession = {
    SparkSession.builder()
      .master("local[*]")
      .appName("AverageByKeyUseReduceByKey")
      .getOrCreate()
  }

  def main(args: Array[String]): Unit = {
    //Create a new spark session
    val spark = createSparkSession
    val input = List(
      ("k1", 1), ("k1", 2), ("k1", 3), ("k1", 4), ("k1", 5),
      ("k2", 6), ("k2", 7), ("k2", 8),
      ("k3", 10), ("k3", 12)
    )
    // build RDD<key, value>
    val rdd = spark.sparkContext.parallelize(input)
    /*
     Combined data structure (C) is a Tuple2(sum, count)
     3 functions needs to be defined:
        v --> C
        C, v --> C
        C, C --> C
    */
    val sumCount = rdd.combineByKey(
      v=> (v,1),
      (C:(Int,Int),v:Int) => (C._1+v , C._2 + 1) ,
      (C1:(Int,Int),C2:(Int,Int)) => (C1._1+C2._1,C1._2+C2._2)
    )
    //show sum count
    println(s"sum_count = ${sumCount.collect().mkString("["," ,","]")}")
    /*
     [
      (k3, (22, 2)),
      (k2, (21, 3)),
      (k1, (15, 5))
     ]
    */

    //find averages
    val avg = sumCount.mapValues(values => (values._1/values._2).toFloat)
    println(s"avg = ${avg.collect().mkString("["," ,","]")}")
    /*
     avg.collect()
     [
      (k3, 11.0),
      (k2, 7.0),
      (k1, 3.0)
     ]
    */
    //Done!
    spark.stop()
  }
}
